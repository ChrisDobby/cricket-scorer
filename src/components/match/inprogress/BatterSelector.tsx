import * as React from 'react';
import * as styles from './styles';

export interface PlayerPosition {
    position: number;
    playerIndex: number;
}

export interface BatterSelectorProps {
    players: string[];
    playerPositions: PlayerPosition[];
    availablePositions: number[];
    notAllowedPlayers?: number[];
    playerSelected: (index: number, position: number) => void;
    playerRemoved: (position: number) => void;
}

const indicatorStyle: React.CSSProperties = {
    float: 'left',
    width: '30px',
    textAlign: 'center',
};

interface PositionIndicatorProps {
    playerIndex: number;
    playerPositions: PlayerPosition[];
}

const PositionIndicator = ({ playerIndex, playerPositions }: PositionIndicatorProps) => {
    const playerWithPosition = playerPositions.find(pp => pp.playerIndex === playerIndex);
    return (
        <div style={indicatorStyle}>
            {playerWithPosition &&
                <span className="badge badge-primary">{playerWithPosition.position}</span>}
        </div>
    );
};

const positionSelected = (
    index: number,
    availablePositions: number[],
    playerPositions: PlayerPosition[],
    playerSelected: (index: number, position: number) => void,
    playerRemoved: (position: number) => void): void => {
    if (availablePositions.length === 1) {
        playerSelected(index, availablePositions[0]);
        return;
    }

    const playerPosition = playerPositions.find(playerPos => playerPos.playerIndex === index);
    if (playerPosition) {
        playerRemoved(playerPosition.position);
        return;
    }

    const filledPositions = playerPositions.map(playerPos => playerPos.position);
    for (let i = 0; i < availablePositions.length; i += 1) {
        if (filledPositions.filter(pos => pos === availablePositions[i]).length === 0) {
            playerSelected(index, availablePositions[i]);
            break;
        }
    }
};

export const BatterSelector = ({
    players, playerPositions, availablePositions, notAllowedPlayers, playerSelected, playerRemoved,
}: BatterSelectorProps) => (
        <div>
            {players.map((player, index) =>
                (
                    <div
                        key={player}
                        className="row"
                        style={ notAllowedPlayers && notAllowedPlayers.find(idx => idx === index)
                            ? styles.nonSelectablePlayerStyle : styles.selectablePlayerStyle}
                        onClick={notAllowedPlayers && notAllowedPlayers.find(idx => idx === index)
                            ? () => { }
                            : () => positionSelected(
                                index,
                                availablePositions,
                                playerPositions,
                                playerSelected,
                                playerRemoved,
                            )}
                    >
                        <PositionIndicator playerIndex={index} playerPositions={playerPositions} />
                        <h6>{player}</h6>
                    </div>
                ))}
        </div>
    );
