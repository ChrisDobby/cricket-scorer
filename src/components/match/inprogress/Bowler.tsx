import * as React from 'react';
import * as styles from './styles';
import * as globalStyles from '../../styles';

const indicatorStyle: React.CSSProperties = {
    ...globalStyles.primaryColour,
    float: 'left',
    width: '30px',
    textAlign: 'center',
};

interface SelectionIndicatorProps {
    playerIndex: number;
    selected: boolean;
}

const SelectionIndicator = ({ playerIndex, selected }: SelectionIndicatorProps) => {
    return (
        <div style={indicatorStyle}>
            {selected && <i className="fa fa-circle" />}
        </div>
    );
};

export interface BowlerProps {
    index: number;
    selected: boolean;
    name: string;
    allowed: boolean;
    selectBowler: (bowler: number) => void;
}

export const Bowler = ({ index, selected, name, allowed, selectBowler }: BowlerProps) => (
    <div
        className="row bowler-row"
        style={allowed ? styles.selectablePlayerStyle : styles.nonSelectablePlayerStyle}
        onClick={() => { if (allowed) { selectBowler(index); } }}
    >
        <SelectionIndicator
            playerIndex={index}
            selected={selected}
        />
        <h6>{name}</h6>
    </div>
);
