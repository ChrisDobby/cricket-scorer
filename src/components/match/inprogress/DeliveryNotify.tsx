import * as React from 'react';
import Snackbar from '@material-ui/core/Snackbar';
import SnackbarContent from '@material-ui/core/SnackbarContent';
import Slide from '@material-ui/core/Slide';
import green from '@material-ui/core/colors/green';
import red from '@material-ui/core/colors/red';
import { notificationDescription } from '../../../match/delivery';
import { Outcome, DeliveryOutcome } from '../../../domain';

interface DeliveryNotifyProps {
    outcome: Outcome;
    onClose: () => void;
}

const notifyStyle: React.CSSProperties = {
    backgroundColor: green[600],
};

const invalidNotifyStyle: React.CSSProperties = {
    backgroundColor: red[700],
};

export default ({ outcome, onClose }: DeliveryNotifyProps) => (
    <Snackbar
        open={true}
        onClose={onClose}
        TransitionComponent={(props: any) => <Slide direction="down" {...props} />}
        anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
        autoHideDuration={3000}
    >
        <SnackbarContent
            style={outcome.deliveryOutcome === DeliveryOutcome.Valid ? notifyStyle : invalidNotifyStyle}
            message={notificationDescription(outcome)}
        />
    </Snackbar>
);
