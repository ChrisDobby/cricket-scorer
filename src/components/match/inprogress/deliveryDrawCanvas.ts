import { Outcome, DeliveryOutcome } from '../../../domain';

const deliveryDrawCanvas = () => {
    const drawDotBall = (context: CanvasRenderingContext2D) => {
        context.beginPath();
        context.arc(15, 15, 4, 0, 2 * Math.PI);
        context.fill();
    };

    const drawScore = (context: CanvasRenderingContext2D, score: number) => context.fillText(score.toString(), 11, 20);

    const drawByes = (context: CanvasRenderingContext2D, byes: number) => {
        context.beginPath();
        drawScore(context, byes);
        context.moveTo(15, 4);
        context.lineTo(2, 24);
        context.lineTo(28, 24);
        context.lineTo(15, 4);
        context.stroke();
    };

    const drawLegByes = (context: CanvasRenderingContext2D, legByes: number) => {
        context.beginPath();
        drawScore(context, legByes);
        context.moveTo(15, 27);
        context.lineTo(2, 8);
        context.lineTo(28, 8);
        context.lineTo(15, 27);
        context.stroke();
    };

    const drawWide = (context: CanvasRenderingContext2D, wides: number) => {
        const dot = (x: number, y: number) => {
            context.beginPath();
            context.arc(x, y, 2, 0, 2 * Math.PI);
            context.fill();
        };

        const dots = [() => dot(8, 8), () => dot(22, 8), () => dot(22, 22), () => dot(8, 22)];

        context.beginPath();
        context.moveTo(15, 2);
        context.lineTo(15, 28);
        context.moveTo(2, 15);
        context.lineTo(28, 15);
        context.stroke();

        for (let i = 0; i < wides; i += 1) {
            dots[i]();
        }
    };

    const drawRuns = (context: CanvasRenderingContext2D, runs: number) => {
        drawScore(context, runs);
    };

    const drawWicket = (context: CanvasRenderingContext2D) => {
        context.fillText('W', 9, 20);
    };

    const drawNoBallOutline = (context: CanvasRenderingContext2D) => {
        context.beginPath();
        context.arc(15, 15, 14, 0, 2 * Math.PI);
        context.stroke();
    };

    const drawOutcome = (canvas: HTMLCanvasElement | undefined, outcome: Outcome) => {
        if (typeof canvas === 'undefined' || canvas === null) {
            return;
        }
        const context = canvas.getContext('2d');
        if (context === null) {
            return;
        }

        context.fillStyle = '#ffffff';
        context.strokeStyle = '#ffffff';
        context.font = '14px Segoe UI';
        context.lineWidth = 1;

        if (typeof outcome.wicket !== 'undefined') {
            drawWicket(context);
        } else if (typeof outcome.scores.byes !== 'undefined') {
            drawByes(context, outcome.scores.byes);
        } else if (typeof outcome.scores.legByes !== 'undefined') {
            drawLegByes(context, outcome.scores.legByes);
        } else if (outcome.deliveryOutcome === DeliveryOutcome.Wide) {
            drawWide(context, typeof outcome.scores.wides === 'undefined' ? 0 : outcome.scores.wides);
        } else if (typeof outcome.scores.runs !== 'undefined' && outcome.scores.runs > 0) {
            drawRuns(context, outcome.scores.runs);
        } else if (typeof outcome.scores.boundaries !== 'undefined') {
            drawRuns(context, outcome.scores.boundaries);
        } else {
            drawDotBall(context);
        }

        if (outcome.deliveryOutcome === DeliveryOutcome.Noball) {
            drawNoBallOutline(context);
        }
    };

    return { drawOutcome };
};

export default deliveryDrawCanvas();
