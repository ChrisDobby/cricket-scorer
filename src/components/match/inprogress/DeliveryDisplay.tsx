import * as React from 'react';
import { Outcome, DeliveryOutcome } from '../../../domain';
import deliveryDrawCanvas from './deliveryDrawCanvas';

const displayStyle: React.CSSProperties = {
    width: '30px',
    height: '30px',
    marginRight: '4px',
    color: '#ffffff',
    display: 'inline-block',
    borderRadius: '4px',
    fontSize: '14px',
};

const validDisplayStyle: React.CSSProperties = {
    ...displayStyle,
    backgroundColor: '#28a745',
};

const noBallDisplayStyle: React.CSSProperties = {
    ...displayStyle,
    backgroundColor: '#dc3545',
};

export interface DeliveryDisplayProps { outcome: Outcome; }

export class DeliveryDisplay extends React.Component<DeliveryDisplayProps, {}> {
    symbolCanvas: HTMLCanvasElement | undefined;

    getCanvas = (canvas: HTMLCanvasElement) => {
        this.symbolCanvas = canvas;
    }

    componentDidMount() {
        deliveryDrawCanvas.drawOutcome(this.symbolCanvas, this.props.outcome);
    }

    render() {
        return (
            <div
                style={this.props.outcome.deliveryOutcome === DeliveryOutcome.Noball ||
                    this.props.outcome.deliveryOutcome === DeliveryOutcome.Wide
                    ? noBallDisplayStyle
                    : validDisplayStyle}
            >
                <canvas ref={this.getCanvas} />
            </div>
        );
    }
}
