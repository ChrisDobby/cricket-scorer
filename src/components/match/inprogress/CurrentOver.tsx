import * as React from 'react';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import { Over } from '../../../domain';
import DeliveryDisplay from './DeliveryDisplay';
import HelpTooltip from '../../HelpTooltip';
import HelpContent from '../../HelpContent';

interface CurrentOverProps {
    over: Over;
}

const overContainerStyle: React.CSSProperties = {
    minHeight: '100px',
    paddingTop: '10px',
    paddingBottom: '10px',
};

export default ({ over }: CurrentOverProps) => (
    <HelpTooltip title={<HelpContent.CurrentOver />}>
        <div style={overContainerStyle}>
            <Grid container>
                <Typography variant="h6">{`This over ${over.wickets} - ${over.bowlingRuns}`}</Typography>
            </Grid>
            <Grid container>
                {over.deliveries.map((delivery, index) => (
                    <DeliveryDisplay key={index} outcome={delivery.outcome} />
                ))}
            </Grid>
        </div>
    </HelpTooltip>
);
