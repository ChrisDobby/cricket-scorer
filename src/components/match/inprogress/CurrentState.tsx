import * as React from 'react';
import Hidden from '@material-ui/core/Hidden';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpandMore from '@material-ui/icons/ExpandMore';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import { Innings, Batter, Bowler, Over, Team } from '../../../domain';
import CurrentOver from './CurrentOver';
import InningsState from './InningsState';

interface CurrentStateProps {
    battingTeam: Team;
    bowlingTeam: Team;
    innings: Innings;
    batter: Batter;
    bowler: Bowler;
    currentOver: Over;
}

export default ({ battingTeam, bowlingTeam, innings, batter, bowler, currentOver }: CurrentStateProps) => (
    <>
        <Hidden xsDown>
            <InningsState
                battingTeam={battingTeam.name}
                battingPlayers={battingTeam.players}
                bowlingPlayers={bowlingTeam.players}
                innings={innings}
                batter={batter}
                bowler={bowler}
            />
        </Hidden>
        <Hidden smUp>
            <ExpansionPanel style={{ width: '100%' }}>
                <ExpansionPanelSummary expandIcon={<ExpandMore />}>
                    <Typography>{`${battingTeam.name} ${innings.score}-${innings.wickets}`}</Typography>
                </ExpansionPanelSummary>
                <ExpansionPanelDetails>
                    <Grid container>
                        <InningsState
                            battingTeam={battingTeam.name}
                            battingPlayers={battingTeam.players}
                            bowlingPlayers={bowlingTeam.players}
                            innings={innings}
                            batter={batter}
                            bowler={bowler}
                        />
                    </Grid>
                </ExpansionPanelDetails>
            </ExpansionPanel>
        </Hidden>
        <CurrentOver over={currentOver} />
    </>
);
