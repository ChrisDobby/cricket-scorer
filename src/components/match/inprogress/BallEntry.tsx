import * as React from 'react';
import Grid from '@material-ui/core/Grid';
import { Innings, Batter, Bowler, Over, Team, MatchResult, DeliveryOutcome, DeliveryScores } from '../../../domain';
import EntryPanel from './EntryPanel';
import DeliveryHeader from '../DeliveryHeader';
import CurrentState from './CurrentState';
import OverCompleteAlert from './OverCompleteAlert';
import EntryContainer from './EntryContainer';

interface BallEntryProps {
    innings: Innings;
    battingTeam: Team;
    bowlingTeam: Team;
    batter: Batter;
    bowler: Bowler;
    overComplete: boolean;
    currentOver: Over;
    delivery: (deliveryOutcome: DeliveryOutcome, scores: DeliveryScores) => void;
    completeOver: () => void;
    homeTeam: string;
    awayTeam: string;
    calculateResult: () => MatchResult | undefined;
}

const entryContainerStyle: React.CSSProperties = {
    marginTop: '10px',
    marginBottom: '10px',
    padding: '5px',
};

export default (props: BallEntryProps) => (
    <Grid container>
        <DeliveryHeader
            batter={props.battingTeam.players[props.batter.playerIndex]}
            bowler={props.bowlingTeam.players[props.bowler.playerIndex]}
            battingPlayers={props.battingTeam.players}
        />
        <Grid container>
            <Grid item xs={12} lg={6} style={entryContainerStyle}>
                <EntryContainer>
                    <CurrentState
                        battingTeam={props.battingTeam}
                        bowlingTeam={props.bowlingTeam}
                        innings={props.innings}
                        batter={props.batter}
                        bowler={props.bowler}
                        currentOver={props.currentOver}
                    />
                </EntryContainer>
            </Grid>
            <Grid item xs={12} lg={6} style={entryContainerStyle}>
                <EntryContainer>
                    <EntryPanel
                        delivery={props.delivery}
                        overComplete={props.overComplete}
                        homeTeam={props.homeTeam}
                        awayTeam={props.awayTeam}
                        calculateResult={props.calculateResult}
                    />
                </EntryContainer>
            </Grid>
        </Grid>
        {props.overComplete && <OverCompleteAlert completeOver={props.completeOver} />}
    </Grid>
);
