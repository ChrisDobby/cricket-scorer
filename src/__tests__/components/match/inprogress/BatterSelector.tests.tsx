import * as React from 'react';
import { shallow } from 'enzyme';
import * as ReactTestRenderer from 'react-test-renderer';
import { BatterSelector } from '../../../../components/match/inprogress/BatterSelector';

describe('BatterSelector', () => {
    const players = [
        'Player 1',
        'Player 2',
        'Player 3',
    ];

    const playerSelected = jest.fn();
    const playerRemoved = jest.fn();

    beforeEach(() => jest.resetAllMocks());

    it('should call playerSelected when an available player is clicked', () => {
        const selector = shallow(
            <BatterSelector
                availablePositions={[1, 2]}
                players={players}
                playerPositions={[]}
                playerSelected={playerSelected}
                playerRemoved={playerRemoved}
            />);

        selector.find('.row').at(0).simulate('click');

        expect(playerSelected).toHaveBeenCalledWith(0, 1);
    });

    it('should call playerSelected for the next position if first is taken', () => {
        const selector = shallow(
            <BatterSelector
                availablePositions={[1, 2]}
                players={players}
                playerPositions={[
                    {
                        position: 1,
                        playerIndex: 0,
                    },
                ]}
                playerSelected={playerSelected}
                playerRemoved={playerRemoved}
            />);

        selector.find('.row').at(1).simulate('click');

        expect(playerSelected).toHaveBeenCalledWith(1, 2);
    });

    it('should call nothing if clicking a new player when all positions are taken', () => {
        const selector = shallow(
            <BatterSelector
                availablePositions={[1, 2]}
                players={players}
                playerPositions={[
                    {
                        position: 1,
                        playerIndex: 0,
                    },
                    {
                        position: 2,
                        playerIndex: 1,
                    },
                ]}
                playerSelected={playerSelected}
                playerRemoved={playerRemoved}
            />);

        selector.find('.row').at(2).simulate('click');

        expect(playerSelected).toHaveBeenCalledTimes(0);
    });

    it('should call playerRemoved when clicking a player with a position', () => {
        const selector = shallow(
            <BatterSelector
                availablePositions={[1, 2]}
                players={players}
                playerPositions={[
                    {
                        position: 1,
                        playerIndex: 0,
                    },
                ]}
                playerSelected={playerSelected}
                playerRemoved={playerRemoved}
            />);

        selector.find('.row').at(0).simulate('click');

        expect(playerRemoved).toHaveBeenCalledWith(1);
    });

    it('should render correctly when no positions filled', () => {
        const selector = ReactTestRenderer.create(
            <BatterSelector
                availablePositions={[1, 2]}
                players={players}
                playerPositions={[]}
                playerSelected={playerSelected}
                playerRemoved={playerRemoved}
            />);

        expect(selector.toJSON()).toMatchSnapshot();
    });

    it('should render correctly when all positions filled', () => {
        const selector = ReactTestRenderer.create(
            <BatterSelector
                availablePositions={[1, 2]}
                players={players}
                playerPositions={[
                    {
                        position: 1,
                        playerIndex: 0,
                    },
                    {
                        position: 2,
                        playerIndex: 1,
                    },
                ]}
                playerSelected={playerSelected}
                playerRemoved={playerRemoved}
            />);

        expect(selector.toJSON()).toMatchSnapshot();
    });
});
