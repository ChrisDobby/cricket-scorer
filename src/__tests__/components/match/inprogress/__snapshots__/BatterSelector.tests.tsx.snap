// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatterSelector should render correctly when all positions filled 1`] = `
<div>
  <div
    className="row"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <div
      style={
        Object {
          "float": "left",
          "textAlign": "center",
          "width": "30px",
        }
      }
    >
      <span
        className="badge badge-primary"
      >
        1
      </span>
    </div>
    <h6>
      Player 1
    </h6>
  </div>
  <div
    className="row"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <div
      style={
        Object {
          "float": "left",
          "textAlign": "center",
          "width": "30px",
        }
      }
    >
      <span
        className="badge badge-primary"
      >
        2
      </span>
    </div>
    <h6>
      Player 2
    </h6>
  </div>
  <div
    className="row"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <div
      style={
        Object {
          "float": "left",
          "textAlign": "center",
          "width": "30px",
        }
      }
    />
    <h6>
      Player 3
    </h6>
  </div>
</div>
`;

exports[`BatterSelector should render correctly when no positions filled 1`] = `
<div>
  <div
    className="row"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <div
      style={
        Object {
          "float": "left",
          "textAlign": "center",
          "width": "30px",
        }
      }
    />
    <h6>
      Player 1
    </h6>
  </div>
  <div
    className="row"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <div
      style={
        Object {
          "float": "left",
          "textAlign": "center",
          "width": "30px",
        }
      }
    />
    <h6>
      Player 2
    </h6>
  </div>
  <div
    className="row"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <div
      style={
        Object {
          "float": "left",
          "textAlign": "center",
          "width": "30px",
        }
      }
    />
    <h6>
      Player 3
    </h6>
  </div>
</div>
`;
